[WGSL] Add a tiny DSL for declaring built-in types

tadeuzagallo opened the pull request:
#### 8375c946e46d51914c60cf86e060b1d6e8765d65
<pre>
[WGSL] Add a tiny DSL for declaring built-in types
<a href="https://bugs.webkit.org/show_bug.cgi?id=252251">https://bugs.webkit.org/show_bug.cgi?id=252251</a>
rdar://105457268

Reviewed by Myles C. Maxfield.

Initializing all the built-in types in C++ is quite verbose, so this patch adds a tiny
embedded DSL in JavaScript for declaring such types.

* Source/WebGPU/WGSL/Overload.h:
* Source/WebGPU/WGSL/TypeCheck.cpp:
(WGSL::TypeChecker::TypeChecker):
* Source/WebGPU/WGSL/TypeDeclarations.rb: Added.
* Source/WebGPU/WGSL/generator/main.rb: Added.
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/260433@main">https://commits.webkit.org/260433@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9cabde058244d1f16f92bccfac458a7a7ffc7b70

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108111 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/17181 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/40979 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/117236 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/116557 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/18702 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8477 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/100315 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/113879 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/14119 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97201 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/41907 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/96026 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/28839 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/83580 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10063 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30187 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10784 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/7089 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16211 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49778 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7217 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12365 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/WebGPU/WGSL/TypeCheck.cpp:
```diff
@@ -118,33 +118,8 @@ TypeChecker::TypeChecker(ShaderModule& shaderModule)
-        WTFMove(plus2),
-    }));
+    // This file contains the declarations generated from `TypeDeclarations.rb`
+#include "TypeDeclarations.h"
```
I think you need a `// NOLINT` here to shut up the style checker.

djg commented on Source/WebGPU/WGSL/TypeDeclarations.js:
```diff
@@ -0,0 +1,6 @@
+// FIXME: add all the missing type declarations here
+type('+', [Number(T)], [T, T], T)
+type('+', [Number(T), N], [Vector(T, N), T], T)
```
Do we need to add support to constrain `C`, `R` and `N` to the values 2, 3 or 4?

tadeuzagallo replied:
My plan was to explicit instantiate e.g. `Vector(T, 2)` instead of constraining the `NumericVariable`s

djg commented on Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
```diff
@@ -617,6 +640,14 @@
 			path = AST;
 			sourceTree = "<group>";
 		};
+		97296763299BFB72001C8BD4 /* generator */ = {
+			isa = PBXGroup;
+			children = (
+				97296767299D33B9001C8BD4 /* main.js */,
+			);
+			path = generator;
+			sourceTree = "<group>";
+		};
```
Instead of running the generator on every build, would it suffice just to update the generated files when we need to change `TypeDeclarations.js`? I don't see the benefit of having everyone run this step all the time when once the WGSL compiler is stable this file won't change very often and isn't much burden for WebGPU devs to update the generated code and submit it.

djg replied:
Also given the issues of running `jsc` in the build causing the failures on EWS.

tadeuzagallo replied:
I'd think it's more convenient to generate at build time, given it only takes <20ms. I'm chatting with folks to try and figure out the right way to invoke JSC at build time. I think adding a manual step is a bit error prone, and I don't love the idea of checking in generated files, but I'm not fundamentally opposed.

litherum replied:
I agree with Tadeu. Let's not check in generated files, if we can help it.

djg reviewed (approved):
:shipit:

litherum commented on Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
```diff
@@ -150,6 +151,24 @@
+				"$(BUILT_PRODUCTS_DIR)/DerivedSources/WGSL/TypeDeclarations.h",
+			);
+			runOncePerArchitecture = 0;
+			script = "/System/Library/Frameworks/JavaScriptCore.framework/Versions/Current/Helpers/jsc \"${SCRIPT_INPUT_FILE_0}\" -- \"${INPUT_FILE_PATH}\" > \"${SCRIPT_OUTPUT_FILE_0}\"\n";
```
Is this going to work on customer machines?

tadeuzagallo replied:
Yes, it passes EWS, which is technically a customer install (also I've had `jsc` aliased to that in my personal computers for as long as I can remember)

tadeuzagallo committed:
[WGSL] Add a tiny DSL for declaring built-in types
https://bugs.webkit.org/show_bug.cgi?id=252251
rdar://105457268

Reviewed by Myles C. Maxfield.

Initializing all the built-in types in C++ is quite verbose, so this patch adds a tiny
embedded DSL in JavaScript for declaring such types.

* Source/WebGPU/WGSL/Overload.h:
* Source/WebGPU/WGSL/TypeCheck.cpp:
(WGSL::TypeChecker::TypeChecker):
* Source/WebGPU/WGSL/TypeDeclarations.rb: Added.
* Source/WebGPU/WGSL/generator/main.rb: Added.
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/260433@main
